src/core/auth/wallet-authenticator.ts

```diff
@@ -0,0 +1,184 @@
+ * @example
+ * ```javascript
+ * // We choose an application name on the server side
+ * const application = "my-app-name"
+ *
+ * // On the server side, we can generate a payload for a wallet requesting to authenticate
+ * const payloadWithSignature = await sdk.auth.generate({
+ *   application,
+ *   subject: "0x...",
+ * })
+ *
+ * // Then on the client side, we can sign this payload with the wallet requesting to authenticate
+ * const authenticatedPayload = await sdk.auth.sign(payloadWithSignature)
+ *
+ * // Finally, on the server side, we can verify if this payload is valid
+ * const isValid = await sdk.auth.verify(application, authenticatedPayload)
+ * ```
+ * @public
+ */
+export class WalletAuthenticator extends RPCConnectionHandler {
+  constructor(network: NetworkOrSignerOrProvider, options: SDKOptions) {
+    super(network, options);
+  }
+
+  public async generate(
+    payload: AuthenticationPayloadInput,
+  ): Promise<AuthorizedPayload> {
+    const parsedPayload: FilledAuthenticationPayloadInput =
+      AuthenticationPayloadInputSchema.parse(payload);
+
+    const signer = this.requireSigner();
+    const signerAddress = await signer.getAddress();
+
+    const payloadOutput: AuthenticationPayloadOutput =
+      AuthenticationPayloadOutputSchema.parse({
+        // Add the issuer in application:address format to use on the client side
+        iss: `${parsedPayload.application}:${signerAddress}`,
+        sub: parsedPayload.subject,
+        aud: parsedPayload.audience,
+        nbf: parsedPayload.invalidBefore,
+        exp: parsedPayload.expiresAt,
+      });
+
+    const payloadMessage = JSON.stringify(payloadOutput);
+    const signature = await signer.signMessage(payloadMessage);
+
+    const authorizedPayload = AuthorizedPayloadSchema.parse({
+      payload: payloadOutput,
+      authorizedPayload: signature,
+    });
+
+    return authorizedPayload;
+  }
+
+  public async sign(
+    authorizedPayload: AuthorizedPayload,
+  ): Promise<AuthenticatedPayload> {
+    // Ensure that admin address is the same as the issuer
+    const adminAddress = this.recoverAddress(
+      JSON.stringify(authorizedPayload.payload),
+      authorizedPayload.authorizedPayload,
+    );
+    if (
+      authorizedPayload.payload.iss.split(":")[1]?.toLowerCase() !==
+      adminAddress.toLowerCase()
+    ) {
+      throw Error(
+        "Payload issuer is not the same as the address that authorized the payload.",
+      );
+    }
+
+    const signerAddress = await this.requireSigner().getAddress();
+    if (
+      authorizedPayload.payload.sub.toLowerCase() !==
+      signerAddress.toLowerCase()
+    ) {
+      throw Error(
+        "Payload subject is not the same as the wallet connected to the SDK.",
+      );
+    }
+
+    const message = this.generateMessage(authorizedPayload);
+    const signer = this.requireSigner();
+    const authenticatedPayload = await signer.signMessage(message);
```

Comment:
how about using the `UserWallet` class here? I expose a bunch of things that you're re-implementing here like `requireSigner` and `signMessage`

you could pass in a `UserWallet` class to the constructor of this class

Reply:
good idea